content/posts/windows-containers.md

```diff
@@ -0,0 +1,231 @@
+  containers:
+  - name: jnlp
+    image: jenkins/jnlp-agent:latest-windows
+    tty: true
+  - name: windows-dotnet
+    image: mcr.microsoft.com/dotnet/core/sdk:2.1
+    tty: true
+  nodeSelector:
+    agentpool: winage
+```
+
+- The ```nodejs-pod.yaml``` has the ```node```(nodeJS) container definition and will use the default Linux based ```jnlp``` (```jenkins/jnlp-slave:3.27-1```) mentioned before.
+
+```
+kind: Pod
+metadata:
+  name: nodejs-app
+spec:
+  containers:
+  - name: nodejs
+    image: node:slim
+    command:
+    - cat
+    tty: true
+  nodeSelector:
+    agentpool: linage
+```
+
+- Notice both pod yaml definitions use  ```nodeSelector``` ([nodeSelector](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/)) to decide where these pods should be scheduled. If this is not specified, the Kubernetes scheduler will provision the pods following the [default behavior](https://kubernetes.io/blog/2017/03/advanced-scheduling-in-kubernetes/) which could possibly start a pod in the a node with the wrong OS. The tags used for the nodeSelectors are the default tags assigned by Azure when specifyng the nodepool name: ``` agentpool : nameOfNodePool ```. To find the node tags you can use the command: ``` kubectl get nodes --show-labels ```. Other options to prevent scheduling errors are ```taints```  ([more info] (https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/)).
+  
+
+
+    [![](/img/windows-containers/error.png)](/img/windows-containers/error.png)*Scheduling error when not using nodeSelectors. Debug using ```kubectl describe podname```*
+  
+
+- Jenkinsfile:
+
+```
+pipeline {
+  agent none
+  options { 
+    buildDiscarder(logRotator(numToKeepStr: '2'))
+    skipDefaultCheckout true
+  }
+  stages {
+    stage('Test-linux') {
+      agent {
+        kubernetes {
+          label 'nodejs-pod'
+          yamlFile 'linux/nodejs-pod.yaml'
+        }
+      }
+      steps {
+        checkout scm
+        container('nodejs') {
+          echo 'Hello World!'   
+          sh 'node --version'
+        }
+      }
+    }
+    stage('Test-windows') {
+      agent {
+        kubernetes {
+	  label 'windows-pod'
+          yamlFile 'windows/dotnet-pod.yaml'
+        }
+      }
+      steps {
+        bat 'dir'
+        container(name:'windows-dotnet'){
+          bat 'dotnet -h'
+      } 
+     }
+    }
+  }
+}
+```
+
+
+- This is a declarative pipeline using agent ```none```([none](https://jenkins.io/doc/book/pipeline/syntax/#agent) info) so that we can specify agents per stage. ```yamlFile``` is used to read the pod template from a file location which is also in this repo. You can either define the pod template in another file, in the same [Jenkinsfile](https://github.com/jenkinsci/kubernetes-plugin/blob/master/examples/declarative-multiple-containers.groovy) or in the Jenkins Configuration page. More info about the syntax used in this pipeline can be found [here](https://jenkins.io/doc/book/pipeline/syntax/). 
```

Comment:
I might simplify here as well:
"This is a declarative pipeline using [`agent none`](https://jenkins.io/doc/book/pipeline/syntax/#agent)"